compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -519,6 +550,152 @@ impl<'a> PartialEvaluator<'a> {
+        // evaluate.
+        if let Value::Bool(condition_bool) = condition_value {
+            let maybe_if_expr_value = if condition_bool {
+                let maybe_body_value = self.try_eval_expr(body_expr_id);
+                if let Ok(body_value) = maybe_body_value {
+                    Ok(body_value)
+                } else {
+                    let body_expr = self.get_expr(body_expr_id);
+                    let error = Error::FailedToEvaluateIfExpressionBranchBlock(body_expr.span);
+                    Err(error)
+                }
+            } else if let Some(otherwise_expr_id) = otherwise_expr_id {
+                let maybe_otherwise_value = self.try_eval_expr(otherwise_expr_id);
+                if let Ok(otherwise_value) = maybe_otherwise_value {
+                    Ok(otherwise_value)
+                } else {
+                    let otherwise_expr = self.get_expr(otherwise_expr_id);
+                    let error = Error::FailedToEvaluateIfExpressionBranchBlock(otherwise_expr.span);
+                    Err(error)
+                }
+            } else {
+                Ok(Value::unit())
+            };
+            return maybe_if_expr_value;
+        }
+
+        // At this point the condition value is not classical, so we need to generate a branching instruction.
+        // First, we pop the current block node and generate a new one which the new branches will jump to when their
+        // instructions end.
+        let current_block_node = self.eval_context.pop_block_node();
+        let continuation_block_node_id = self.create_program_block();
+        let continuation_block_node = BlockNode {
+            id: continuation_block_node_id,
+            next: current_block_node.next,
+        };
+        self.eval_context.push_block_node(continuation_block_node);
+
+        // Since the if expression can represent a dynamic value, create a variable to store it if the expression is
+        // non-unit.
+        let if_expr = self.get_expr(if_expr_id);
+        let maybe_if_expr_var = if if_expr.ty == Ty::UNIT {
+            None
+        } else {
+            let variable_id = self.resource_manager.next_var();
+            let variable_ty = map_fir_type_to_rir_type(&if_expr.ty);
+            Some(Variable {
+                variable_id,
+                ty: variable_ty,
+            })
+        };
+
+        // Evaluate the body expression.
+        let (body_block_node_id, body_expr_value) =
+            self.eval_expr_if_branch(body_expr_id, continuation_block_node_id)?;
+        // If there is a variable to save to, add a store instruction at the end of the body block.
+        if let Some(if_expr_var) = maybe_if_expr_var {
+            let body_operand = map_eval_value_to_rir_operand(&body_expr_value);
+            let store_ins = Instruction::Store(body_operand, if_expr_var);
+            let body_block = self.get_program_block_mut(body_block_node_id);
+            body_block.0.push(store_ins);
```

Comment:
This is a bug because this store instruction will happen after the jump instruction.

Reply:
Add test that fails.